lib/Dialect/Calyx/CalyxOps.cpp

```diff
@@ -2135,6 +2189,264 @@ LogicalResult WhileOp::canonicalize(WhileOp whileOp,
+  result.addAttribute("inputNames",
+                      ArrayAttr::get(parser.getContext(), inputNames));
+  return success();
+}
+
+void InvokeOp::print(OpAsmPrinter &p) {
+  // print the parameter list.
+  p << " @" << getCallee() << "(";
+  auto ports = this->getPorts();
+  auto inputs = this->getInputs();
+  for (size_t i = 0; i != ports.size(); i++) {
+    p << ports[i] << " = " << inputs[i];
+    if (i + 1 != ports.size())
+      p << ", ";
+  }
+  p << ") -> (";
+  // Print argument type.
+  for (size_t i = 0; i != ports.size(); i++) {
+    p << ports[i].getType();
+    if (i + 1 != ports.size())
+      p << ", ";
+  }
+  p << ")";
+}
+
+static LogicalResult verifyInvokeOpValue(InvokeOp op, Value value,
+                                         bool isDestination) {
+  bool isSource = !isDestination;
+  if (isPort(value))
+    return verifyPortDirection(op, value, isDestination);
+
+  if (isSource) {
+    // Refer to the above function verifyNotComplexSource for its role.
+    Operation *operation = value.getDefiningOp();
+    if (operation == nullptr)
+      return success();
+    if (auto *dialect = operation->getDialect();
+        isa<comb::CombDialect>(dialect))
+      return op->emitOpError("has source that is not a port or constant. "
+                             "Complex logic should be conducted in the guard.");
+  }
+  return success();
+}
+
+// Get the value of the Go port within an instance.
+Value InvokeOp::getInstGoValue() {
+  ComponentOp componentOp = (*this)->getParentOfType<ComponentOp>();
+  Operation *operation = componentOp.lookupSymbol(getCallee());
+  Value ret = nullptr;
+  // Get the value of the writer_en port.
+  if (isa<RegisterOp>(operation))
+    ret = operation->getResult(1);
+  else if (isa<MemoryOp, DivSPipeLibOp, DivUPipeLibOp, MultPipeLibOp,
+               RemSPipeLibOp, RemUPipeLibOp>(operation)) {
+    // Get the value of the writer_en port or go port.
+    ret = operation->getResult(2);
+  } else if (isa<InstanceOp>(operation)) {
+    // Get the go port of the instance through the "go" attribute
+    InstanceOp instanceOp = cast<InstanceOp>(operation);
+    auto portInfo = instanceOp.getReferencedComponent().getPortInfo();
+    for (size_t i = 0; i != portInfo.size(); i++) {
+      if (portInfo[i].hasAttribute("go"))
+        ret = operation->getResult(i);
+    }
+  } else if (isa<PrimitiveOp>(operation)) {
+    // Get the go part of the primitive through the "calyx.go" attribute
+    PrimitiveOp primOp = cast<PrimitiveOp>(operation);
+    auto moduleExternOp = primOp.getReferencedPrimitive();
+    auto argAttrs = moduleExternOp.getArgAttrsAttr();
+    for (size_t i = 0; i != argAttrs.size(); i++)
+      if (DictionaryAttr dictAttr = dyn_cast<DictionaryAttr>(argAttrs[i]))
+        if (!dictAttr.empty())
+          if (dictAttr.begin()->getName().getValue() == "calyx.go")
+            ret = primOp.getResult(i);
+  }
+
+  return ret;
```

Comment:
Can we return some error in case one of these cases isn't covered? Right now it seems to return `nullptr`.